sql/auth/sql_user_table.cc

```diff
@@ -2026,15 +2038,39 @@ int open_grant_tables(THD *thd, TABLE_LIST *tables,
     Perform the TOI after the replication filter check to avoid
     replicating commands that won't be applied locally (due to a filter).
   */
-  /* Doing TOI here is not the ideal solution, as we are holding
-  MDL locks already and the current thread can still be BF-aborted.
-  But we have to check replication filters before TOI and for this
-  we need MDL locks. */
   if (WSREP(thd) && !skip_toi &&
       wsrep_to_isolation_begin(thd, db, table, nullptr)) {
     WSREP_ERROR("Fail to replicate: %s", thd->query().str);
     return -1;
```

Comment:
We need to call `thd->mdl_context.release_transactional_locks();` here as well.
```suggestion
    WSREP_ERROR("Fail to replicate: %s", thd->query().str);
    thd->mdl_context.release_transactional_locks();
    return -1;
```

Reply:
Well, we are not holding locks during TOI replication so no unlocking if it fails.

Reply:
Thanks for clarifying. Got it! 👍🏼